### PR TITLE
Account Store implementation

### PR DESCRIPTION
**Pull request by tonyr59h**
This PR implements the Account Store by mirroring existing Account related functionality in WPAndroid.

**TODO:**
- [x] Handle errors in `AccountStore`
- [x] Discuss unimplemented methods (unmarked below)
- [x] Unit tests
- [x] Regression tests

### Relevant Account related classes in WPAndroid
These are the existing classes and their relevant methods. I've marked the methods that have an analogue in this PR. Unmarked methods are ones I did not think needed to be implemented, feel free to comment on them though :)
* **AccountModel** - basic data model to store Account attributes
	- [x] `init`: sets attribute default values, ensures non-null object values
	- [x] `updateFromRestResponse`: updates `/me` related attributes
	- [x] `updateAccountSettingsFromRestResponse`: updates `/me/settings` related attributes
	- [x] `RestParam`: enum containing `/me/settings` response keys
* **Account** - an extension of `AccountModel` that provides convenience methods for REST calls
	- [x] `fetchAccountDetails`: performs HTTP GET on the `/me` endpoint
	- [x] `fetchAccountSettings`: performs HTTP GET on the `/me/settings` endpoint
	- [x] `postAccountSettings`: performs HTTP POST on the `/me/settings` endpoint
* **AccountTable** - provides SQL database convenience methods
	- [x] `createTables`: creates the Account table
	- [x] `dropTables`: drops the Account table
	- [x] `save`: saves the given Account to the Account table
	- [x] `getDefaultAccount`: retrieves the Account with `localId=0` from the Account table
	- [x] `getAccountByLocalId`: retrieves an Account by a given `localId` from the Account table
* **AccountHelper** - provides convenience methods common Account queries
	- [x] `getDefaultAccount`: retrieves the Account with `localId=0` and stores as singleton for future reference
	- [x] `isSignedIn`: checks if the default Account is signed in
	- [x] `isSignedInWordPressDotCom`: checks if the default Account is signed into WordPress.com
	- [x] `isJetPackUser`: checks if the current Account has any blogs with Jetpack enabled
	- [x] `getCurrentUsernameForBlog`: gets the username from the default Account or given `Blog`

#### AccountModel
Simple data model object that replaces most of `AccountModel` from WPAndroid and provides some SQL functionality that existed in `AccountTable` via WellSql annotations.

##### Methods converted:

WPAndroid | WPStores
--- | --- | ---
AccountTable.createTables | WellSqlConfig.onCreate + `@Table` annotation
AccountTable.dropTables | WellSqlConfig.reset + `@Table` annotation
AccountModel.init | init
AccountModel.updateFromRestResponse | copyAccountAttributes
AccountModel.updateAccountSettingsFromRestResponse | copyAccountSettingsAttributes

#### AccountStore
Handles AccountAction events, initiates WordPress.com Account related REST calls, authenticates Account credentials, handles Account related network errors, and persists Account changes in a SQL database.

##### Methods converted:

WPAndroid | WPStores
--- | ---
AccountHelper.getDefaultAccount | getAccount
AccountHelper.isSignedInWordPressDotCom | hasAccessToken
AccountHelper.isSignedIn | isSignedIn

##### Actions handled:

AccountAction | Description
--- | ---
**AUTHENTICATE_ERROR** | emits an `OnAuthenticationChanged` event with the authentication error
**AUTHENTICATE** | attempts to authenticate credentials from a given payload
**FETCH** | fetches Account and Account Settings data via REST API
**FETCH_ACCOUNT** | fetches only Account data via REST API
**FETCH_ACCOUNT_SETTINGS** | fetches only Account Settings data via REST API
**FETCHED_ACCOUNT** | updates Account information from a `/me` endpoint response and persists it in the SQL database
**FETCHED_SETTINGS** | updates Account Settings information from a `/me/settings response and persists it in the SQL database
**POST_SETTINGS** | posts Account Settings changes remotely via REST API
**POSTED_SETTINGS** | checks for error
**UPDATE** | sets the in-memory Account to the Account passed in the payload then persists it to the SQL database

Since Account and Account Settings endpoint responses are asynchronous I've modified the `FETCH` action to run both Account endpoint GET calls immediately. ~~This required some state tracking that you'll notice with the `mFetchStatus` field and `moveFetchStatus` method. Subsequent `FETCH` actions are ignored until the in-progress `FETCH` action gets a response from both endpoints.~~

#### AccountRestClient
Provides methods to interface with Account related calls to the WordPress.com REST API.

##### Methods converted:

WPAndroid | WPStores | Success response
--- | --- | --- | ---
Account.fetchAccountDetails | fetchAccount | **FETCHED_ACCOUNT**
Account.fetchAccountSettings | fetchAccountSettings | **FETCHED_SETTINGS**
Account.postAccountSettings | postAccountSettings | **POSTED_SETTINGS**

Each of the two REST endpoints returns unique data so I've added unique objects to handle responses from each, `AccountResponse` and `AccountSettingsResponse`. The `RestParams` enum from WPAndroid is analogous to the field names in the response classes.

I've renamed the existing `getMe` method to `fetchAccount` to stay semantically consistent with other method names (fetch == network calls).

#### AccountSqlUtils
Provides SQL database convenience methods using WellSql, replacing most of the `AccountTable` functionality from WPAndroid. Account and Account Settings are given distinct methods since REST responses are asynchronous.

WPAndroid | WPStores
--- | ---
AccountTable.save | insertOrUpdateAccount / updateAccount
AccountTable.getDefaultAccount | getDefaultAccount
AccountTable.getAccountByLocalId | getAccountByLocalId

Additional helper methods provided: `deleteAccount`, `getAllAccounts`

### CODE REVIEW

**Comment by maxme** [comment]
> `isSignedIn`: checks if the default Account is signed in

It's used to test if a user has .org sites or is signed in a .com account. (if not `isSignedIn`: show the login screen)

File: example/src/androidTest/java/org/wordpress/android/stores/release/ReleaseStack_AccountTest.java [hunk 1]
```diff
@@ -34,36 +43,52 @@ protected void setUp() throws Exception {
...
+    }
 
-        // Correct user we should get an OnAuthenticationChanged message
-        mDispatcher.dispatch(AuthenticationAction.AUTHENTICATE, payload);
-        mCountDownLatch = new CountDownLatch(1);
-        // Wait for a network response / onChanged event
-        assertEquals(true, mCountDownLatch.await(TestUtils.DEFAULT_TIMEOUT_MS, TimeUnit.MILLISECONDS));
+    public void testWPCOMAuthenticationError() throws InterruptedException {
+        mExpectedAction = ACCOUNT_TEST_ACTIONS.AUTHENTICATE_ERROR;
+        authenticate(BuildConfig.TEST_WPCOM_USERNAME_TEST1, BuildConfig.TEST_WPCOM_BAD_PASSWORD);
```
Comment (aforcier) [inline example/src/androidTest/java/org/wordpress/android/stores/release/ReleaseStack_AccountTest.java]: `TEST_WPCOM_BAD_PASSWORD` should be added to `tests.properties-example`.

**Comment by aforcier** [comment]
> `isJetPackUser`: checks if the current Account has any blogs with Jetpack enabled

This is implemented in SiteStore as `hasJetpackSite()`. It's `WordPressDB.hasAnyJetpackBlogs()` on WPAndroid, which is what `AccountHelper.isJetPackUser()` is calling, but it's actually querrying the blogs table.

As this information would rely on `/me/sites/` anyway which is handled by the SiteStore anyway, it seems to me like we can ignore this method for `AccountStore`. What do you think @maxme?

**Comment by aforcier** [comment]
> `getCurrentUsernameForBlog`: gets the username from the default Account or given `Blog`

The way I see it, this can be dropped. In the two instances the method is used in WPAndroid, one could be replaced by `AccountStore.getAccount().getUserName()` and the other by `SiteModel.getUserName()`. The last one is tricky since it depends on `currentBlog`, but we're axing that by passing around a reference to a `SiteModel`, so we wouldn't need the `AccountStore` reference in that case anyway.

**Comment by maxme** [comment]
> As this information would rely on /me/sites/ anyway which is handled by the SiteStore anyway, it seems to me like we can ignore this method for AccountStore. What do you think @maxme?

Yep, we should ignore this in AccountStore. A quick grep in wpandroid shown `isJetPackUser` is only used in Analytics, I guess we would ask the SiteStore when we init analytics and set the isJetpack flag then.

**Comment by maxme** [comment]
> The way I see it, this can be dropped. In the two instances the method is used in WPAndroid, one could be replaced by AccountStore.getAccount().getUserName() and the other by SiteModel.getUserName(). The last one is tricky since it depends on currentBlog, but we're axing that by passing around a reference to a SiteModel, so we wouldn't need the AccountStore reference in that case anyway.

Exactly, `WordPress.getCurrentBlog()` should die with wpstores integration in wpandroid.

**Comment by tonyr59h** [comment]
Thanks for the feedback, I agree on the the two methods not being necessary.

The last REST call test was just added in 8033659f6fd2dd9c18766660a06840a97bff277e, I think this is ready for merge now.

**Comment by aforcier** [comment]
Looks good! :shipit: as far as I'm concerned, letting @maxme weigh in also.

**Comment by maxme** [comment]
2FA is lacking (App + SMS code), but I'll just fill a new issue for this.

**Comment by maxme** [comment]
:shipit:

**Pull request merged**